# Data Mining Project - Phase 6 - Ehsan Espandar - 99442011

<div lang='rtl'>
    با سلام، به دلیل پیشنیاز بودن داده های پیش پردازش شده ی فاز سه و چهار آنها نیز در این نوتبوک آورده شده اند. لطفا به سکشن فاز 6 برای بررسی فایل مراجعه کنید
</div>

In [48]:
import pandas as pd

In [49]:
data_packet = pd.read_csv('housing.csv')

In [50]:
from sklearn.model_selection import train_test_split

X = data_packet.drop('median_house_value', axis=1)
y = data_packet['median_house_value']


X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2021)

print(len(X_train),len(y_train))
print(len(X_test),len(y_test))

16512 16512
4128 4128


## Phase 4 (It is repeated because phase 5 needs the information of label and prepared data.)

In [51]:
import copy
import pandas as pd

# Assuming X_train and y_train are your feature and target DataFrames respectively
X_train_copy = copy.copy(X_train)
y_train_copy = copy.copy(y_train)

# Concatenate X_train and y_train along columns axis
data = pd.concat([X_train_copy, y_train_copy], axis=1)


print(data)

       longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
2782     -114.66     32.74                17.0       1388.0           386.0   
8513     -118.31     33.89                35.0       2144.0           423.0   
2163     -119.82     36.77                41.0       1441.0           274.0   
5232     -118.24     33.94                42.0        380.0           106.0   
9766     -121.66     36.37                 9.0       1580.0           287.0   
...          ...       ...                 ...          ...             ...   
16347    -121.35     38.02                16.0       1665.0           311.0   
20318    -119.15     34.20                25.0       3445.0           898.0   
2669     -115.51     33.24                32.0       1995.0           523.0   
17536    -121.88     37.35                49.0       1728.0           350.0   
6201     -117.89     34.08                25.0       2115.0           489.0   

       population  households  median_income ocean_

In [52]:
label = data["median_house_value"].copy()

data_without_label = data.drop("median_house_value", axis=1)

# Display the copied feature and the dataset without "ocean_proximity"
print("Copied label 'median_house_value':\n", label)
print("\nDataset without 'median_house_value':\n", data_without_label)



Copied label 'median_house_value':
 2782      44000.0
8513     231500.0
2163      71300.0
5232      90000.0
9766     405800.0
           ...   
16347    132300.0
20318    169300.0
2669      43300.0
17536    193000.0
6201     207400.0
Name: median_house_value, Length: 16512, dtype: float64

Dataset without 'median_house_value':
        longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
2782     -114.66     32.74                17.0       1388.0           386.0   
8513     -118.31     33.89                35.0       2144.0           423.0   
2163     -119.82     36.77                41.0       1441.0           274.0   
5232     -118.24     33.94                42.0        380.0           106.0   
9766     -121.66     36.37                 9.0       1580.0           287.0   
...          ...       ...                 ...          ...             ...   
16347    -121.35     38.02                16.0       1665.0           311.0   
20318    -119.15     34.20            

In [53]:
# Create a copy of the "ocean_proximity" feature
ocean_proximity_copy = data_without_label["ocean_proximity"].copy()

# Remove the "ocean_proximity" feature from the original dataset
data_without_ocean_proximity = data_without_label.drop("ocean_proximity", axis=1)

# Display the copied feature and the dataset without "ocean_proximity"
print("Copied feature 'ocean_proximity':\n", ocean_proximity_copy)
print("\nDataset without 'ocean_proximity':\n", data_without_ocean_proximity)


Copied feature 'ocean_proximity':
 2782         INLAND
8513      <1H OCEAN
2163         INLAND
5232      <1H OCEAN
9766      <1H OCEAN
            ...    
16347        INLAND
20318    NEAR OCEAN
2669         INLAND
17536     <1H OCEAN
6201      <1H OCEAN
Name: ocean_proximity, Length: 16512, dtype: object

Dataset without 'ocean_proximity':
        longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
2782     -114.66     32.74                17.0       1388.0           386.0   
8513     -118.31     33.89                35.0       2144.0           423.0   
2163     -119.82     36.77                41.0       1441.0           274.0   
5232     -118.24     33.94                42.0        380.0           106.0   
9766     -121.66     36.37                 9.0       1580.0           287.0   
...          ...       ...                 ...          ...             ...   
16347    -121.35     38.02                16.0       1665.0           311.0   
20318    -119.15     34.

In [54]:
from sklearn.impute import SimpleImputer

#As data has a normal distribution and not heavy amount of outliers then the 'mean' strategy is the method we use to normalize the dataset.

# Selecting only numerical features
numerical_features = ['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income']

# Creating a SimpleImputer instance with a chosen strategy (e.g., mean)
imputer = SimpleImputer(strategy='mean')

# Fitting the imputer on the data and transforming it to fill missing values
data_without_ocean_proximity[numerical_features] = imputer.fit_transform(data_without_ocean_proximity[numerical_features])

print(data_without_ocean_proximity)



       longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
2782     -114.66     32.74                17.0       1388.0           386.0   
8513     -118.31     33.89                35.0       2144.0           423.0   
2163     -119.82     36.77                41.0       1441.0           274.0   
5232     -118.24     33.94                42.0        380.0           106.0   
9766     -121.66     36.37                 9.0       1580.0           287.0   
...          ...       ...                 ...          ...             ...   
16347    -121.35     38.02                16.0       1665.0           311.0   
20318    -119.15     34.20                25.0       3445.0           898.0   
2669     -115.51     33.24                32.0       1995.0           523.0   
17536    -121.88     37.35                49.0       1728.0           350.0   
6201     -117.89     34.08                25.0       2115.0           489.0   

       population  households  median_income  
2782

In [55]:
imputed_data_df = pd.DataFrame(data_without_ocean_proximity, columns=numerical_features)
import numpy as np
from sklearn.base import BaseEstimator, TransformerMixin
rooms_ix,bedrooms_ix,population_ix,household_ix=3,4,5,6

class CombinedAttributesAdder(BaseEstimator,TransformerMixin):
    def fit(self,x,y=None):
        return self
    def transform(self,x,y=None):
        rooms_per_household=x[:,rooms_ix]/x[:,household_ix]
        population_per_household=x[:,population_ix]/x[:,household_ix]
        bedrooms_per_room=x[:,bedrooms_ix]/x[:,rooms_ix]
        return np.c_[x,rooms_per_household,population_per_household,bedrooms_per_room]

custom=CombinedAttributesAdder()
data_custom_tr_top=custom.transform(imputed_data_df.values)
data_custom_tr=pd.DataFrame(data_custom_tr_top)
columns=list(imputed_data_df)
columns.append('rooms_per_household')
columns.append('population_per_household')
columns.append('bedrooms_per_room')
data_custom_tr.columns=columns
data_custom_tr.head(10)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -114.66     32.74                17.0       1388.0           386.0   
1    -118.31     33.89                35.0       2144.0           423.0   
2    -119.82     36.77                41.0       1441.0           274.0   
3    -118.24     33.94                42.0        380.0           106.0   
4    -121.66     36.37                 9.0       1580.0           287.0   
5    -121.85     37.36                15.0       3148.0          1116.0   
6    -122.49     37.79                52.0       2488.0           281.0   
7    -117.87     33.76                 6.0       2992.0          1194.0   
8    -118.22     33.88                35.0        998.0           313.0   
9    -122.20     39.15                33.0       1064.0           174.0   

   population  households  median_income  rooms_per_household  \
0       775.0       320.0         1.2049             4.337500   
1      1192.0       417.0         4.1458             5.141487   
2       646.0       296.0         3.0568             4.868243   
3       411.0       100.0         0.9705             3.800000   
4       465.0       208.0         6.1668             7.596154   
5      3556.0      1037.0         3.0466             3.035680   
6       805.0       295.0        10.7058             8.433898   
7      3800.0      1130.0         2.2460             2.647788   
8      1335.0       311.0         1.6574             3.209003   
9       434.0       147.0         3.1250             7.238095   

   population_per_household  bedrooms_per_room  
0                  2.421875           0.278098  
1                  2.858513           0.197295  
2                  2.182432           0.190146  
3                  4.110000           0.278947  
4                  2.235577           0.181646  
5                  3.429122           0.354511  
6                  2.728814           0.112942  
7                  3.362832           0.399064  
8                  4.292605           0.313627  
9                  2.952381           0.163534

In [56]:
from sklearn.preprocessing import StandardScaler

# Selecting only numerical features
numerical_features = ['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'rooms_per_household', 'population_per_household', 'bedrooms_per_room']

# Create a StandardScaler instance
scaler = StandardScaler()

# Fit the scaler on the data and transform it to standardize the features
data_custom_tr[numerical_features] = scaler.fit_transform(data_custom_tr[numerical_features])

print(data_custom_tr)

       longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0       2.453114 -1.356402           -0.919431    -0.567531       -0.361595   
1       0.630571 -0.818013            0.507393    -0.224185       -0.274180   
2      -0.123413  0.530301            0.983001    -0.543460       -0.626200   
3       0.665524 -0.794605            1.062269    -1.025325       -1.023109   
4      -1.042174  0.343035           -1.553576    -0.480332       -0.595487   
...          ...       ...                 ...          ...             ...   
16507  -0.887383  1.115506           -0.998699    -0.441728       -0.538786   
16508   0.211136 -0.672882           -0.285287     0.366678        0.848031   
16509   2.028686 -1.122320            0.269589    -0.291855       -0.037925   
16510  -1.152026  0.801836            1.617146    -0.413116       -0.446646   
16511   0.840288 -0.729061           -0.285287    -0.237356       -0.118252   

       population  households  median_income  rooms

In [57]:
from sklearn.preprocessing import LabelEncoder

# Create a LabelEncoder instance
label_encoder = LabelEncoder()

# Fit and transform the 'ocean_proximity' column
ocean_proximity_copy = label_encoder.fit_transform(ocean_proximity_copy)
print(ocean_proximity_copy)


[1 0 1 ... 1 0 0]


In [58]:
from sklearn.preprocessing import OneHotEncoder
import pandas as pd

# Assuming 'ocean_proximity_copy' is your encoded column as a NumPy array
ocean_proximity_copy = ocean_proximity_copy.reshape(-1, 1)  # Reshape to a 2D array

# Create a OneHotEncoder
onehot_encoder = OneHotEncoder()

# Fit and transform the specified column using the OneHotEncoder
ocean_proximity_onehot = onehot_encoder.fit_transform(ocean_proximity_copy)

# Extract the dense representation as an array
ocean_proximity_onehot_array = ocean_proximity_onehot.toarray()

# Convert to a DataFrame
ocean_proximity_onehot_df = pd.DataFrame(ocean_proximity_onehot_array, columns=["ocean_proximity_0", "ocean_proximity_1", "ocean_proximity_2", "ocean_proximity_3", "ocean_proximity_4"])

# Find the index of the maximum value along the columns (axis=1)
ocean_proximity_encoded = ocean_proximity_onehot_df.idxmax(axis=1)

# Extract the numeric part from the column names
ocean_proximity_encoded = ocean_proximity_encoded.str.extract(r'(\d+)')

# Rename the column
ocean_proximity_encoded.columns = ["ocean_proximity"]

# Convert the result to integers
ocean_proximity_encoded = ocean_proximity_encoded.astype(int)

# Display the resulting Series
print(ocean_proximity_encoded)


       ocean_proximity
0                    1
1                    0
2                    1
3                    0
4                    0
...                ...
16507                1
16508                4
16509                1
16510                0
16511                0

[16512 rows x 1 columns]


In [59]:
# Assuming 'data_custom_tr' is your original DataFrame and 'ocean_proximity_onehot_df' is the one-hot encoded DataFrame
data_custom_tr = pd.concat([data_custom_tr, ocean_proximity_encoded], axis=1)


# Display the resulting DataFrame
print(data_custom_tr)


       longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0       2.453114 -1.356402           -0.919431    -0.567531       -0.361595   
1       0.630571 -0.818013            0.507393    -0.224185       -0.274180   
2      -0.123413  0.530301            0.983001    -0.543460       -0.626200   
3       0.665524 -0.794605            1.062269    -1.025325       -1.023109   
4      -1.042174  0.343035           -1.553576    -0.480332       -0.595487   
...          ...       ...                 ...          ...             ...   
16507  -0.887383  1.115506           -0.998699    -0.441728       -0.538786   
16508   0.211136 -0.672882           -0.285287     0.366678        0.848031   
16509   2.028686 -1.122320            0.269589    -0.291855       -0.037925   
16510  -1.152026  0.801836            1.617146    -0.413116       -0.446646   
16511   0.840288 -0.729061           -0.285287    -0.237356       -0.118252   

       population  households  median_income  rooms

# **Phase 6**







---



# Question 1

In [60]:
from sklearn.model_selection import GridSearchCV

# Question 2

In [61]:
param_grid = [{'n_estimators': [3,4,6,10,30],"max_features":[2,6,8,15]}]

# Question 3

In [62]:
from sklearn.ensemble import RandomForestRegressor

forest_reg = RandomForestRegressor()

# Question 4

In [63]:
grid_search = GridSearchCV(forest_reg, param_grid, cv=5, scoring='neg_mean_squared_error')

# Question 5

In [64]:
housing_prepared = data_custom_tr

grid_search.fit(housing_prepared,label)

GridSearchCV(cv=5, estimator=RandomForestRegressor(),
             param_grid=[{'max_features': [2, 6, 8, 15],
                          'n_estimators': [3, 4, 6, 10, 30]}],
             scoring='neg_mean_squared_error')

# Question 6

In [65]:
print("Best Parameters:",grid_search.best_params_)
print("Best Estimator:",grid_search.best_estimator_)

Best Parameters: {'max_features': 6, 'n_estimators': 30}
Best Estimator: RandomForestRegressor(max_features=6, n_estimators=30)


In [66]:
results=grid_search.cv_results_
print(results)

{'mean_fit_time': array([0.10313444, 0.13393488, 0.19527888, 0.32224717, 1.13421736,
       0.25735917, 0.30949469, 0.45534444, 0.75642633, 2.47589002,
       0.2967093 , 0.49803123, 0.6942904 , 0.97591653, 3.12618999,
       0.5403758 , 0.68768387, 0.85497618, 1.6344873 , 4.44740582]), 'std_fit_time': array([0.01180398, 0.00914208, 0.00504204, 0.0107992 , 0.24285171,
       0.03882848, 0.00815668, 0.00979545, 0.01169955, 0.27553053,
       0.0095715 , 0.09050062, 0.11848975, 0.01619386, 0.43336873,
       0.0857424 , 0.12450111, 0.01610954, 0.25584073, 0.4613414 ]), 'mean_score_time': array([0.00645442, 0.00754094, 0.01056571, 0.01452222, 0.04773889,
       0.00672321, 0.00893092, 0.01183057, 0.0156601 , 0.04554009,
       0.0065423 , 0.00967441, 0.01068983, 0.0173471 , 0.04404416,
       0.00914683, 0.00927792, 0.01156096, 0.01715384, 0.04395938]), 'std_score_time': array([0.00069559, 0.00029867, 0.00088936, 0.00013813, 0.00709319,
       0.00114386, 0.00078368, 0.00147939, 0.0012310

# Question 7

In [67]:
for mean_score,params in zip(results['mean_test_score'],results["params"]):
  print(np.sqrt(-mean_score),params)

64003.80640943929 {'max_features': 2, 'n_estimators': 3}
59896.79124839691 {'max_features': 2, 'n_estimators': 4}
56799.61920039817 {'max_features': 2, 'n_estimators': 6}
54634.19835275994 {'max_features': 2, 'n_estimators': 10}
51738.08909844773 {'max_features': 2, 'n_estimators': 30}
59697.29914455299 {'max_features': 6, 'n_estimators': 3}
56730.99160340931 {'max_features': 6, 'n_estimators': 4}
54327.46640635574 {'max_features': 6, 'n_estimators': 6}
52639.04501112476 {'max_features': 6, 'n_estimators': 10}
50213.216859778375 {'max_features': 6, 'n_estimators': 30}
59000.50732979773 {'max_features': 8, 'n_estimators': 3}
56756.36901471981 {'max_features': 8, 'n_estimators': 4}
54798.96199937883 {'max_features': 8, 'n_estimators': 6}
52451.15313581154 {'max_features': 8, 'n_estimators': 10}
50670.53724418404 {'max_features': 8, 'n_estimators': 30}
59078.864912725534 {'max_features': 15, 'n_estimators': 3}
57383.271756669616 {'max_features': 15, 'n_estimators': 4}
55511.26554572335 {'

# Question 8 and 9

<div dir="rtl">
از آنجا که نوشتن پایپ لاین در فاز 4 اختیاری بود و من ننوشته ام در مرحله ی  8 همه ی کار هایی که بر روی داده آموزش انجام داده ام برای داده ی تست نیز انجام میدهم.
</div>

In [79]:

# Assuming X_test is your test dataset
X_test_copy = copy.copy(X_test)

# Extract the 'ocean_proximity' feature
ocean_proximity_copy_test = X_test_copy['ocean_proximity'].copy()

# Remove the 'ocean_proximity' feature from the original dataset
data_without_ocean_proximity_test = X_test_copy.drop('ocean_proximity', axis=1)

# Selecting only numerical features
numerical_features_test = ['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income']

# Creating a SimpleImputer instance with the strategy as mean
imputer_test = SimpleImputer(strategy='mean')

# Fitting the imputer on the test data and transforming it to fill missing values
data_without_ocean_proximity_test[numerical_features_test] = imputer_test.fit_transform(data_without_ocean_proximity_test[numerical_features_test])

# Creating an instance of CombinedAttributesAdder
custom_test = CombinedAttributesAdder()

# Transforming the imputed test data using CombinedAttributesAdder
data_custom_tr_top_test = custom_test.transform(data_without_ocean_proximity_test.values)
data_custom_tr_test = pd.DataFrame(data_custom_tr_top_test)

# Adding columns to the DataFrame
columns_test = list(data_without_ocean_proximity_test)
columns_test.append('rooms_per_household')
columns_test.append('population_per_household')
columns_test.append('bedrooms_per_room')
data_custom_tr_test.columns = columns_test

# Selecting only numerical features
numerical_features_test = ['longitude', 'latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms', 'population', 'households', 'median_income', 'rooms_per_household', 'population_per_household', 'bedrooms_per_room']

# Create a StandardScaler instance
scaler_test = StandardScaler()

# Fit the scaler on the test data and transform it to standardize the features
data_custom_tr_test[numerical_features_test] = scaler_test.fit_transform(data_custom_tr_test[numerical_features_test])

# Create a LabelEncoder instance
label_encoder = LabelEncoder()

# Fit and transform the 'ocean_proximity' column
ocean_proximity_copy_encoded = label_encoder.fit_transform(ocean_proximity_copy_test)

# Reshape to a 2D array
ocean_proximity_copy_encoded = ocean_proximity_copy_encoded.reshape(-1, 1)

# Create a OneHotEncoder
onehot_encoder = OneHotEncoder()
ocean_proximity_onehot = onehot_encoder.fit_transform(ocean_proximity_copy_encoded)

# Convert to a DataFrame
ocean_proximity_onehot_df = pd.DataFrame(ocean_proximity_onehot.toarray(), columns=["ocean_proximity_0", "ocean_proximity_1", "ocean_proximity_2", "ocean_proximity_3", "ocean_proximity_4"])

# Find the index of the maximum value along the columns (axis=1)
ocean_proximity_copy_encoded = ocean_proximity_onehot_df.idxmax(axis=1)

# Extract the numeric part from the column names
ocean_proximity_copy_encoded = ocean_proximity_copy_encoded.str.extract(r'(\d+)')

# Rename the column
ocean_proximity_copy_encoded.columns = ["ocean_proximity"]

# Convert the result to integers
ocean_proximity_copy_encoded = ocean_proximity_copy_encoded.astype(int)

# Assuming 'data_custom_tr' is your original DataFrame and 'ocean_proximity_onehot_df' is the one-hot encoded DataFrame
data_custom_tr_test = pd.concat([data_custom_tr_test, ocean_proximity_copy_encoded], axis=1)



In [80]:
final_model = grid_search.best_estimator_

# We have the transformed test data below now as X_prepared and test_labels as Y
X_prepared = data_custom_tr_test
Y = y_test

print(X_prepared)



      longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0     -1.342007  1.165986            1.861474    -1.134358       -1.110194   
1     -0.883515  1.119155            0.256661     0.879576        0.982852   
2     -0.155909  0.552506           -1.749355    -0.331834       -0.283404   
3      1.179697 -1.339447            0.336902    -0.721182       -0.539138   
4      0.626517 -0.852410           -0.625986    -0.743580       -0.621072   
...         ...       ...                 ...          ...             ...   
4123   0.721205 -0.740016            0.577624    -0.541043       -0.539138   
4124   0.905599 -0.735333           -0.144542    -0.196968       -0.231264   
4125  -1.122728  0.763243           -0.625986     0.296269        0.190821   
4126  -1.436696  1.011445            1.861474     0.036545        0.282687   
4127   0.686320 -0.814945            0.738105     0.017006        0.230547   

      population  households  median_income  rooms_per_househol

# Question 10

In [94]:
final_predictions = final_model.predict(X_prepared)

print(final_predictions)

[196170.13333333 123693.33333333  94240.         ... 213580.
 313803.4        199386.66666667]


# Question 11

In [95]:
from sklearn.metrics import mean_squared_error


final_mse= mean_squared_error(Y,final_predictions)
final_rmse=np.sqrt(final_mse)
final_rmse

52256.85081136011

# Question 12

<div dir="rtl">
مدل نهایی با مقدار خطای میانگین 52256.85081136011 دلار قابلیت پیشبینی قیمت خانه هارا دارد

</div>